bootstrap/eks/controllers/eksconfig_controller.go

```diff
@@ -0,0 +1,234 @@
+	if apierrors.IsNotFound(err) {
+		// no error here, requeue until we find an owner
+		return ctrl.Result{}, nil
+	}
+	if err != nil {
+		log.Error(err, "Failed to get owner")
+		return ctrl.Result{}, err
+	}
+	if configOwner == nil {
+		// no error, requeue until we find an owner
+		return ctrl.Result{}, nil
+	}
+
+	log = log.WithValues("kind", configOwner.GetKind(), "version", configOwner.GetResourceVersion(), "name", configOwner.GetName())
+
+	cluster, err := util.GetClusterByName(ctx, r.Client, configOwner.GetNamespace(), configOwner.ClusterName())
+	if err != nil {
+		if errors.Cause(err) == util.ErrNoCluster {
+			log.Info(fmt.Sprintf("%s does not belong to a cluster yet, requeueing until it's part of a cluster", configOwner.GetKind()))
+			return ctrl.Result{}, nil
+		}
+		if apierrors.IsNotFound(err) {
+			log.Info("Cluster does not exist yet, requeueing until it is created")
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, "Could not get cluster with metadata")
+		return ctrl.Result{}, err
+	}
+
+	if annotations.IsPaused(cluster, config) {
+		log.Info("Reconciliation is paused for this object")
+		return ctrl.Result{}, nil
+	}
+
+	scope := &EKSConfigScope{
+		Logger:      log,
+		Config:      config,
+		ConfigOwner: configOwner,
+		Cluster:     cluster,
+	}
+
+	patchHelper, err := patch.NewHelper(config, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// set up defer block for updating config
+	defer func() {
+		conditions.SetSummary(config,
+			conditions.WithConditions(
+				bootstrapv1.DataSecretAvailableCondition,
+			),
+			conditions.WithStepCounter(),
+		)
+
+		patchOpts := []patch.Option{}
+		if rerr == nil {
+			patchOpts = append(patchOpts, patch.WithStatusObservedGeneration{})
+		}
+		if err := patchHelper.Patch(ctx, config, patchOpts...); err != nil {
+			log.Error(rerr, "Failed to patch config")
+			if rerr == nil {
+				rerr = err
+			}
+		}
+	}()
+
+	return r.joinWorker(ctx, scope)
+}
+
+func (r *EKSConfigReconciler) joinWorker(ctx context.Context, scope *EKSConfigScope) (ctrl.Result, error) {
+
+	if !scope.Cluster.Status.InfrastructureReady {
+		scope.Logger.Info("Cluster infrastructure is not ready, requeueing")
+		conditions.MarkFalse(scope.Config,
+			bootstrapv1.DataSecretAvailableCondition,
+			bootstrapv1.WaitingForClusterInfrastructureReason,
+			clusterv1.ConditionSeverityInfo, "")
+		return ctrl.Result{}, nil
+	}
+
+	if !scope.Cluster.Status.ControlPlaneInitialized {
+		scope.Logger.Info("Cluster has not yet been initialized, requeueing")
+		conditions.MarkFalse(scope.Config, bootstrapv1.DataSecretAvailableCondition, bootstrapv1.DataSecretGenerationFailedReason, clusterv1.ConditionSeverityWarning, "")
+		return ctrl.Result{}, nil
+	}
+
+	scope.Logger.Info("generating userdata", "cluster", scope.Cluster.GetName())
+
+	// generate userdata
+	userDataScript, err := userdata.NewNode(&userdata.NodeInput{
+		ClusterName:      scope.Cluster.GetName(),
```

Comment:
I'm assuming that this relates to the underlying EKS Cluster name, if so, do we expect it to actually match the CAPI Cluster Name?

It might be good to come up with a convention here to not get into yet another situation where we could have conflicts on Cluster Name across namespaces....

Reply:
Oof, great point. This value needs to be the EKS cluster name from AWS' perspective - meaning it technically may not be the Cluster object's name unless we enforce that as a standard. 

Curious to get @richardcase's take on this as well.

Reply:
At the moment the cluster name is the same. I could suffix it with -eks to follow the conventions used for other resources?

Reply:
We could then add a func to the managed scope to get the underlying EKS cluster name which i will set in the status of the AWSManagedControlPlane. What do you think?

Reply:
I think going with the Cluster's name is fine. Just wanted to confirm!

Reply:
Still worried about namespace clashes.

Comparing the acceptable characters of both:

AWS cluster name regex: ^[0-9A-Za-z][A-Za-z0-9\-_]*
Kubernetes: RFC1123
* contain no more than 253 characters
* contain only lowercase alphanumeric characters, '-' or '.'
* start with an alphanumeric character
* end with an alphanumeric character

That means someone could create a cluster with name "foo.bar" and this won't work.

Can I recommend that we use the following convention:

EKS cluster name: `<namespace>_<name>`, and with `.` also replaced with `_`?

Reply:
There is also a 100 character limit for EKS, vs the 253 for k8s, so possibly also something similar to what we do for the ELB ( https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/master/pkg/cloud/services/elb/loadbalancer.go#L262 )

Reply:
Good point about namespace clashes. I'll change it like you mention about and hash if its over the 100 character limit.